contracts/eosio.system/src/finalizer_key.cpp

```diff
@@ -0,0 +1,158 @@
+      finalizer_policy fin_policy {
+         .generation = 0;
+         .threshold  = 15;
+         .finalizers = finalizers;
+      };
+      set_finalizers(std::move(fin_policy));
+
+      _gstate5.last_finalizers = finalizers;
+
+      return true;
+   }
+
+   void replace_key_in_finalizer_policy(const name& producer, const std::string& finalizer_key) {
+      // replace the key in last finalizer policy and call set_finalizers
+      // host function immediately
+   }
+
+   void delete_key_in_finalizer_policy(const name& producer, const std::string& finalizer_key) {
+      // replace the key in last finalizer policy and call set_finalizers
+      // host function immediately
+   }
+
+   void system_contract::switchtosvnn() {
+      require_auth(get_self());
+
+      check(!is_savanna_consensus(), "switchtosvnn can only be run once");
+
+      // Find top 21 block producers
+      auto idx = _producers.get_index<"prototalvote"_n>();
+      std::vector<name> top_producers;
+      top_producers.reserve(21);
+      for( auto it = idx.cbegin(); it != idx.cend() && top_producers.size() < 21 && 0 < it->total_votes && it->active(); ++it ) {
+         top_producers.emplace_back(it->owner);
+      }
+      if( top_producers.size() < 21 ) {
+         return;
+      }
+
+      set_next_finalizers_set(top_producers);
+   }
+
+   void system_contract::regfinkey( const name& producer, const std::string& finalizer_key, const std::string& proof_of_possession) {
+      require_auth( producer );
+
+      check( is_registered_producer(producer), "The producer is not a registered producer");
+
+      // Basic key and signature format check
+      check(finalizer_key.compare(0, 7, "PUB_BLS") == 0, "finalizer key must start with PUB_BLS");
+      check(proof_of_possession.compare(0, 7, "SIG_BLS") == 0, "proof of possession siganture must start with SIG_BLS");
+
+      // Proof of possession check
+      const auto pk = eosio::decode_bls_public_key_to_g1(finalizer_key);
+      const auto signature = eosio::decode_bls_signature_to_g2(proof_of_possession);
+      check(eosio::bls_pop_verify(pk, signature), "proof of possession check failed");
+
+      // Todo: Go over finalizer_keys_table to make sure no duplicate keys are
+      // registered.
+
+      auto it = _finalizer_keys.find( producer.value );
+      if( it != _finalizer_keys.end() ) {
+         _finalizer_keys.modify( it, same_payer, [&]( auto& k ) {
+            k.registered_finalizer_keys.insert(finalizer_key);
+         });
+      } else {
+         // The first time the producer registers a finalizer key
+         _finalizer_keys.emplace( producer, [&]( auto& p ) {
+            k.producer = producer;
+            k.registered_finalizer_keys.insert(finalizer_key);
+            k.active_key = finalizer_key;
+         });
+
+         if( is_active_producer(producer) ) {
+            replace_key_in_finalizer_policy(producer, finalizer_key);
```

Comment:
If this is the first time a producer is registering a key, they do not get to benefit from the fast path. They have to wait for `update_elected_producers`.

The fast path is only reserved for an active BP changing their active finalizer key.